File: crawl.test.js
const { normalizeURL, getURLsFromHTML } = require("./crawl.js")
const {test,expect} = require('@jest/globals')

//normalizeURL takes a URL input and returns a normalized URL string. 
//ex: it should remove the protocol (http:// or https://) and any trailing slashes from the URL. It should also convert the URL to lowercase.

test('normalizeURL strip protocol', () => {
    const input = 'https://example.com/path'
    const actual = normalizeURL(input)
    const expected = 'example.com/path'
    
    expect(actual).toEqual(expected)

})

test('normalizeURL strip trailing slash', () => {
    const input = 'https://example.com/path/'
    const actual = normalizeURL(input)
    const expected = 'example.com/path'

    expect(actual).toEqual(expected)

})

test('normalizeURL strip capitals', () => {
    const input = 'https://EXAMPLE.COM/path/'
    const actual = normalizeURL(input)
    const expected = 'example.com/path'

    expect(actual).toEqual(expected)

})

test('normalizeURL strip http', () => {
    const input = 'http://example.com/path/'
    const actual = normalizeURL(input)
    const expected = 'example.com/path'

    expect(actual).toEqual(expected)

})

//getURLsFromHTML takes an HTML body and a baseURL as input and returns an array of all the URLs found in the HTML body. 
//ex: if the HTML body contains <a href="http://example.com/path/">Link</a>, then the function should return ['example.com/path'].
test('getURLsFromHTML absolute', () => {
    const inputHTMLBody = `
<html>
    <body>
        <a href="https://example.com"> 
            example.com 
        </a>
    </body>
</html>
`
    const inputBaseURL = "https://example.com"
    const actual = getURLsFromHTML(inputHTMLBody, inputBaseURL)
    const expected = ["https://example.com/"]
    expect(actual).toEqual(expected)

})

test('getURLsFromHTML relative', () => {
    const inputHTMLBody = `
<html>
    <body>
        <a href="/path/"> 
            example.com 
        </a>
    </body>
</html>
`
    const inputBaseURL = "https://example.com"
    const actual = getURLsFromHTML(inputHTMLBody, inputBaseURL)
    const expected = ["https://example.com/path/"]
    expect(actual).toEqual(expected)

})

test('getURLsFromHTML both', () => {
    const inputHTMLBody = `
<html>
    <body>
        <a href="https://example.com/path1/"> 
            example.com Path One
        </a>
        <a href="/path2/"> 
            example.com Path Two
        </a>
    </body>
</html>
`
    const inputBaseURL = "https://example.com"
    const actual = getURLsFromHTML(inputHTMLBody, inputBaseURL)
    const expected = ["https://example.com/path1/", "https://example.com/path2/"]
    expect(actual).toEqual(expected)

})


test('getURLsFromHTML Invalid', () => {
    const inputHTMLBody = `
<html>
    <body>
        <a href="invalid"> 
            Invalid URL
        </a>
    </body>
</html>
`
    const inputBaseURL = "https://example.com"
    const actual = getURLsFromHTML(inputHTMLBody, inputBaseURL)
    const expected = []
    expect(actual).toEqual(expected)

})

File: crawlTest.py
import crawl

def test_getURLsFromHTML_Absolute():
    #parser = crawl.MyHTMLParser()
    print("testing getURLsFromHTML absolute")
    inputHTML = '''
    <html>
        <body>
            <a href="https://blog.boot.dev/path/"> 
                Boot.dev Blog
            </a>
        </body>
    </html>
    '''

    inputBaseURL = "https://blog.boot.dev/path/"
    actual = crawl.getURLsFromHTML(inputHTML, inputBaseURL)
    expected = ["https://blog.boot.dev/path/"]
    print("Actual: " + str(actual))
    print("Expected: " + str(expected))
    assert actual == expected

def test_getURLsFromHTML_Relative():
    print("testing getURLsFromHTML relative")
    inputHTML = '''
    <html>
        <body>
            <a href="/path/"> 
                Boot.dev Blog
            </a>
        </body>
    </html>
    '''

    inputBaseURL = "https://blog.boot.dev"
    actual = crawl.getURLsFromHTML(inputHTML, inputBaseURL)
    expected = ["https://blog.boot.dev/path/"]
    print("Actual: " + str(actual))
    print("Expected: " + str(expected))
    assert actual == expected

def test_getURLsFromHTML_Both():
    print("testing getURLsFromHTML both")
    inputHTML = '''
    <html>
        <body>
            <a href="https://blog.boot.dev/path1/">
                Boot.dev Blog Path One
            </a>
            <a href="/path2/"> 
                Boot.dev Blog Path Two
            </a>
        </body>
    </html>
    '''

    inputBaseURL = "https://blog.boot.dev"
    actual = crawl.getURLsFromHTML(inputHTML, inputBaseURL)
    expected = ["https://blog.boot.dev/path1/", "https://blog.boot.dev/path2/"]
    print("Actual: " + str(actual))
    print("Expected: " + str(expected))
    assert actual == expected

def test_getURLsFromHTML_invalid():
    print("testing getURLsFromHTML invalid")
    inputHTML = '''
    <html>
        <body>
            <a href="invalid"> 
                Invalid URL
            </a>
        </body>
    </html>
    '''

    inputBaseURL = "https://blog.boot.dev"
    actual = crawl.getURLsFromHTML(inputHTML, inputBaseURL)
    expected = []
    print("Actual: " + str(actual))
    print("Expected: " + str(expected))
    assert actual == expected

def test_normalizeURL():
    print("Testing normalizeURL strip protocol")
    assert crawl.normalizeURL("https://www.example.com/path/") == "example.com/path"

    print("Testing normalizeURL strip trailing slashes")
    assert crawl.normalizeURL("https://www.example.com/path/") == "example.com/path"
    
    print("Testing normalizeURL strip capitals")
    assert crawl.normalizeURL("https://EXAMPLE.COM/path/") == "example.com/path"

    
if __name__ == "__main__":
    
    test_normalizeURL()
    test_getURLsFromHTML_Absolute()
    test_getURLsFromHTML_Relative()
    test_getURLsFromHTML_Both()
    test_getURLsFromHTML_invalid()